### ISSUE TITLE
Native crash related to GC?

### ISSUE THREAD

**Issue by jiro-san** [mono/mono#18344]
Hello, we encountered a crash that is similar to what was reported in the issues below:

https://github.com/mono/mono/issues/14364
https://github.com/mono/mono/issues/14200

It looks like the following line is causing the crash:
https://github.com/mono/mono/blob/090e25ec2da954575b4faeb8ac129867011d13b2/mono/sgen/sgen-alloc.c#L409

Right now our build environment is using VS2017 + Xamarin 9.1.7.0, which is the latest available in VS2017. We cannot upgrade the build environment to VS2019 because we target older Android OS (Android 4.2, API Level 17) and support for these OS was dropped in latest Xamarin version. We can run the application when building with VS2019 but there are many bugs, like files not being properly listed etc, that make the application looks too unstable. 

So our crash reported exactly:
01-01 12:01:38.845 F/        ( 3141): * Assertion at /Users/builder/jenkins/workspace/xamarin-android-d15-9/xamarin-android/external/mono/mono/sgen/sgen-alloc.c:378, condition `*p == NULL' not met

The line doesn't match because our version of Xamarin.Android is older.

See below for the full log:
[log_github.txt](https://github.com/mono/mono/files/4025744/log_github.txt)

Unfortunately it happened only once, and we were not able to reproduce it, so I cannot provide a repro-step. 

However since similar issues seem to pop every now and then I suspect there is something in the runtime. Especially because we never encountered this issue in many years when working with Xamarin 6.0.1, and we had it right after upgrading to Xamarin 9.1.7.0.

I noticed that in the reported issue [#14364](https://github.com/mono/mono/issues/14364) peppy is talking about copying some font textures and we see reference to opentk in the logs.

Our app is also heavily doing 3D work and loading textures during boot.
Maybe it might be related?

Is there anyway to work around this issue, or any chance it would get fixed for VS2017?

**Comment by BrzVlad** [comment]
@jiro-san As I mentioned on one of those issues, this crash is most likely a sign that your application does some buffer overflows, which sounds plausible given your scenario (which might use unsafe code). You could try running with debug option `MONO_GC_DEBUG=nursery-canaries`, which might catch the corruption earlier and print information about the corrupted object.

There is no chance to address this issue without a reproduction case. If the debug option would print the object that is overflowed, it might make it possible to figure out where the corruption is, by narrowing down the culprit code.

**Comment by jiro-san** [comment]
@BrzVlad As I mentionned, this app has been deployed on customer devices for many years and as far as I know we never had the issue when using Xamarin 6.0.1, this is why I suspected more the runtime. But yes I will try doing as you say and hope we can reproduce and get more information.

**Comment by jiro-san** [comment]
Hello, tried to run with MONO_GC_DEBUG=nursery-canaries but crash reporting out of memory.
[logcat_canaries.txt](https://github.com/mono/mono/files/4034652/logcat_canaries.txt)

Our app has maybe around 100-200 MB available usually when running. Is there a big overhead when enabling canaries?
I'll try running this in a VM or another HW that has more memory...

**Comment by jiro-san** [comment]
Enabling canaries crash also in VirtualBox (Android x86 6.0) and on our other HW.
[logcat_canaries2_Kitkat.txt](https://github.com/mono/mono/files/4034701/logcat_canaries2_Kitkat.txt)

/Users/builder/jenkins/workspace/xamarin-android-d15-9/xamarin-android/external/mono/mono/mini/mini-codegen.c:1273, condition `reginfo [sreg].born_in > 0' not met

ouch...

**Comment by BrzVlad** [comment]
Could you try also with `no-managed-allocator` ? (`MONO_GC_DEBUG=nursery-canaries,no-managed-allocator`). Looks like something bitrotted in that debug option.

**Comment by jiro-san** [comment]
wow, thanks for the responsiveness! Trying out now...

**Comment by BrzVlad** [comment]
It is plausible that you are unlucky and your application deadlocks during nursery canaries logging (which happens during GC and not in an async safe manner), if your application already does some other logging and it is interrupted by the GC. I wanted in the past to make the logging async safe and I could figure something out, but it will take a little investigation to handle everything and would only happen on master.

It is plausible that the application will get to start if you stop all your logging, which might be unreasonable, depending on your project.

**Comment by jiro-san** [comment]
Thanks for the feedback.
I changed some logging (not all) and it didn't freeze at the end of boot.
Now the endurance test is running on our most recent hardware (Android 4.4). 
To be continued...

**Comment by jiro-san** [comment]
Ouch, freeze again after less than 10 minutes of endurance test. 
I'll remove more logs (not a lot left, but still logging some stuff)...

**Comment by jiro-san** [comment]
It looks stable after removing all logs. The test has been running for 1 hour now, but no special report yet from canaries. I'll leave running it as long as possible post here if there is any news.

**Comment by BrzVlad** [comment]
Ideally you should look for reproducing it first and then try it with canaries enabled. Maybe, for whatever reason, there is no overflow going on in your current configuration, in which case there will be no crash and no canaries log.

**Comment by jiro-san** [comment]
Yes, and maybe for whatever reason there is an error detected with the canaries but the app freezes when it tries to log the canary error.
True that we should be able to consistently reproduce to investigate, but actually here is the story:
- Our stability requirement is that app should be able to run without crash and reboot for 5 days without user operation.
- We had this kind of random crash for years when using Xamarin, up to Xamarin 6.0.4.
- Some of the crashes were on our side, somes were problems in the runtime but we found some workaround, for some we never had a solution.
- From what we know, we never had random crash like this with Xamarin 6.0.4.
- When we tried to upgrade to other versions of Xamarin we always had some issue preventing it: either a noticeable performance drop, some random crash, some other issue.
- Because of .NET Standard we need to upgrade to our environment to VS2017 + Xamarin.Android 9.1.7.0; Actually we validated that it was stable on our Kitkat configuration.
- Because we need to support Android 4.2, we cannot upgrade our environment to VS2019 + Xamarin.Android 10.1.1.0: there are obvious issues when using this environment. Maybe we could do it, but since support for Android 4.2 was dropped in VS2019 we believe this is not a solution.
